readhomer_atlas/library/schema.py

```diff
@@ -191,156 +222,92 @@ def reference_filter(self, queryset, name, value):
+        _, ref = value.rsplit(":", maxsplit=1)
+        predicate = self._build_predicate(queryset, ref, max_rank)
+        return filter_via_ref_predicate(self, queryset, predicate)
 
 
-class PassageLineConnection(Connection):
+class AbstractTextPartNode(DjangoObjectType):
+    label = String()
+    name = String()
     metadata = generic.GenericScalar()
 
     class Meta:
         abstract = True
 
-    @staticmethod
-    def generate_passage_urn(version, object_list):
-        first = object_list[0]
-        last = object_list[-1]
-
-        if first == last:
-            return first.get("urn")
-        line_refs = [line.get("ref") for line in [first, last]]
-        passage_ref = "-".join(line_refs)
-        return f"{version.urn}{passage_ref}"
-
-    def get_ancestor_metadata(self, obj):
-        data = []
-        if obj:
-            data.append(
-                {
-                    # @@@ proper name for this is ref or position?
-                    "ref": obj.ref,
-                    "urn": obj.book.urn,
-                }
-            )
-        return data
-
-    def get_sibling_metadata(self, version, all_queryset, start_idx, count):
-        data = {}
-
-        chunker = get_chunker(
-            all_queryset, start_idx, count, queryset_values=["idx", "urn", "ref"]
+    @classmethod
+    def __init_subclass_with_meta__(cls, **meta_options):
+        meta_options.update(
+            {
+                "model": TextPart,
+                "interfaces": (relay.Node,),
+                "filterset_class": TextPartFilterSet,
+            }
         )
-        previous_objects, next_objects = chunker.get_prev_next_boundaries()
-
-        if previous_objects:
-            data["previous"] = self.generate_passage_urn(version, previous_objects)
-
-        if next_objects:
-            data["next"] = self.generate_passage_urn(version, next_objects)
-        return data
+        super().__init_subclass_with_meta__(**meta_options)
 
-    def get_children_metadata(self, lines_queryset):
-        data = []
-        for line in lines_queryset.values("position", "urn"):
-            data.append(
-                {
-                    # @@@ proper name is lsb or position
-                    "lsb": str(line.get("position")),
-                    "urn": line.get("urn"),
-                }
-            )
-        return data
+    def resolve_metadata(obj, *args, **kwargs):
+        return camelize(obj.metadata)
 
-    def resolve_metadata(self, info, *args, **kwargs):
-        # @@@ resolve metadata.siblings|ancestors|children individually
-        passage_dict = info.context.passage
-        if not passage_dict:
-            return
 
-        urn = passage_dict["urn"]
-        version = passage_dict["version"]
-        lines_queryset = passage_dict["lines_qs"]
+class VersionNode(AbstractTextPartNode):
+    @classmethod
+    def get_queryset(cls, queryset, info):
+        return queryset.filter(kind="version")
```

Comment:
another clear candidate for refactoring post ![](https://github.trello.services/images/mini-trello-icon.png) [Build or re-use a Python library for URN manipulation](https://trello.com/c/rzaNKWCi) (and likely we can add these GQL "Aliases" for text group and work down the line as well in ![](https://github.trello.services/images/mini-trello-icon.png) [Add text group and work endpoints to ATLAS GQL schema](https://trello.com/c/8vzw7Z6B)